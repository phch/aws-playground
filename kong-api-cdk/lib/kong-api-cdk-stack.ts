import * as cdk from '@aws-cdk/core';
import * as cfn_inc from '@aws-cdk/cloudformation-include';
import { KeyPair } from 'cdk-ec2-key-pair';
import * as path from 'path';

export class KongApiCdkStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    
    // Create EC2 keypair for Kong instances and store public/private key in Secrets Manager
    const key = new KeyPair(this, 'KongEc2KeyPair', {
      name: 'KongEc2KeyPair',
      description: 'Key pair for Kong EC2 instances',
      storePublicKey: true,
    });

    // Import Kong CloudFormation template with PostgresDB
    const kongCfn = new cfn_inc.CfnInclude(this, 'KongPostgresTemplate', {
      templateFile: path.join(__dirname, 'kong-postgres.template.yml'),
      preserveLogicalIds: true,
      parameters: {
        'KongVersion': '', // Left blank so it installs latest version
        'KongKeyName': key.keyPairName,
        'KongFleetDesiredSize': '2',
        'KongInstanceType': 't3.medium',
        'KongConfigs': 'KONG_LOG_LEVEL=debug^KONG_ADMIN_LISTEN="0.0.0.0:8001, 0.0.0.0:8444 ssl"',
        'DBClass': 'db.t3.medium',
        'DBHost': '', // Left blank so it creates a new DB instance
      },
    });
    kongCfn.node.addDependency(key);
  }
}
