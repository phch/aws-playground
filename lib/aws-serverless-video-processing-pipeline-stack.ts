import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as codepipeline_actions from '@aws-cdk/aws-codepipeline-actions';
import { Construct, SecretValue, Stack, StackProps } from '@aws-cdk/core';
import {CdkPipeline, ShellScriptAction, SimpleSynthAction} from "@aws-cdk/pipelines";
import { AwsServerlessVideoProcessingStage } from './aws-serverless-video-procesing-stage';

/**
 * The stack that defines the application pipeline
 */
export class AwsServerlessVideoProcessingPipelineStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const sourceArtifact = new codepipeline.Artifact();
        const cloudAssemblyArtifact = new codepipeline.Artifact();

        const pipeline = new CdkPipeline(this, 'Pipeline', {
            // The pipeline name
            pipelineName: 'AwsServerlessVideoProcessingPipeline',
            cloudAssemblyArtifact,

            // Where the source can be found
            sourceAction: new codepipeline_actions.GitHubSourceAction({
                actionName: 'GitHub',
                output: sourceArtifact,
                oauthToken: SecretValue.secretsManager('github-token'),
                owner: 'phch',
                repo: 'aws-serverless-video-processing',
                branch: 'main',
            }),

            // How it will be built and synthesized
            synthAction: SimpleSynthAction.standardNpmSynth({
                sourceArtifact,
                cloudAssemblyArtifact,

                // We need a build step to compile the TypeScript Lambda
                buildCommand: 'npm run build'
            }),
        });

        // This is where we add the application stages
        const preprod = new AwsServerlessVideoProcessingStage(this, 'PreProd');
        const preprodStage = pipeline.addApplicationStage(preprod);
        preprodStage.addActions(new ShellScriptAction({
            actionName: 'TestService',
            useOutputs: {
                ENDPOINT_URL: pipeline.stackOutput(preprod.urlOutput),
            },
            commands: [
                // Use 'curl' to GET the given URL and fail if it returns an error
                'curl -Ssf $ENDPOINT_URL',
            ],
        }));

        pipeline.addApplicationStage(new AwsServerlessVideoProcessingStage(this, 'Prod'));

    }
}