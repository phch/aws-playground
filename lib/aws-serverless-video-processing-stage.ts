import { CfnOutput, Construct, Stage, StageProps } from '@aws-cdk/core';
import { AwsServerlessVideoProcessingStack } from './aws-serverless-video-processing-stack';

/**
 * Deployable unit of web service app
 */
export class AwsServerlessVideoProcessingStage extends Stage {
    public readonly urlOutput: CfnOutput;

    constructor(scope: Construct, id: string, props?: StageProps) {
        super(scope, id, props);

        const service = new AwsServerlessVideoProcessingStack(this, 'WebService');

        // Expose CdkpipelinesDemoStack's output one level higher
        this.urlOutput = service.urlOutput;
    }
}