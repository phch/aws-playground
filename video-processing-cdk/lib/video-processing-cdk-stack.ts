import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as s3 from '@aws-cdk/aws-s3';
import * as s3n from '@aws-cdk/aws-s3-notifications';

export class VideoProcessingCdkStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const inputBucket = new s3.Bucket(this, 'InputVideoBucket');
    const outputBucket = new s3.Bucket(this, 'OutputVideoBucket');

    const layer = new lambda.LayerVersion(this, 'FfmpegLayer', {
      description: 'Lambda layer for ffmpeg',
      code: lambda.Code.fromAsset('lambda/layer/ffmpeg-4.4.zip'),
      compatibleRuntimes: [lambda.Runtime.PYTHON_3_8],
      license: 'GPLv3',
    });

    const fn = new lambda.Function(this, 'VideoProcessingFn', {
      code: lambda.Code.fromAsset('lambda'),
      handler: 'index.handler',
      runtime: lambda.Runtime.PYTHON_3_8,
      layers: [layer],
      environment: {
        'S3_DESTINATION_BUCKET': outputBucket.bucketName
      }
    });
    inputBucket.addEventNotification(s3.EventType.OBJECT_CREATED_PUT, new s3n.LambdaDestination(fn));
    inputBucket.grantRead(fn);
    outputBucket.grantWrite(fn);
  }
}
