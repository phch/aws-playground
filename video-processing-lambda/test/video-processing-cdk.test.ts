import { expect as expectCDK, countResources, haveResource, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as VideoProcessingCdk from '../lib/video-processing-cdk-stack';

test('Contains required resources', () => {
    const app = new cdk.App();
    // WHEN
    const stack = new VideoProcessingCdk.VideoProcessingCdkStack(app, 'TestStack');
    // THEN
    expectCDK(stack).to(haveResource('AWS::Lambda::LayerVersion'));
    expectCDK(stack).to(haveResource('AWS::Lambda::Function'));
    expectCDK(stack).to(countResources('AWS::S3::Bucket', 2));
});

test('Matches snapshot', () => {
  const app = new cdk.App();
  // WHEN
  const stack = new VideoProcessingCdk.VideoProcessingCdkStack(app, 'TestStack');
  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
