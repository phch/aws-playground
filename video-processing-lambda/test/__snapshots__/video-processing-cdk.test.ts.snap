// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters819e1c802f7a3ee062b88ee47bec19d4f4f9c0e9475221ce3d678e85581305a5ArtifactHashE978C27E": Object {
      "Description": "Artifact hash for asset \\"819e1c802f7a3ee062b88ee47bec19d4f4f9c0e9475221ce3d678e85581305a5\\"",
      "Type": "String",
    },
    "AssetParameters819e1c802f7a3ee062b88ee47bec19d4f4f9c0e9475221ce3d678e85581305a5S3Bucket181207DA": Object {
      "Description": "S3 bucket for asset \\"819e1c802f7a3ee062b88ee47bec19d4f4f9c0e9475221ce3d678e85581305a5\\"",
      "Type": "String",
    },
    "AssetParameters819e1c802f7a3ee062b88ee47bec19d4f4f9c0e9475221ce3d678e85581305a5S3VersionKey842E1070": Object {
      "Description": "S3 key for asset version \\"819e1c802f7a3ee062b88ee47bec19d4f4f9c0e9475221ce3d678e85581305a5\\"",
      "Type": "String",
    },
    "AssetParametersac7f5d00801594308615088ac65f2f0848f8cc59edd9fc76134593bebef2de31ArtifactHashEBE1FBBA": Object {
      "Description": "Artifact hash for asset \\"ac7f5d00801594308615088ac65f2f0848f8cc59edd9fc76134593bebef2de31\\"",
      "Type": "String",
    },
    "AssetParametersac7f5d00801594308615088ac65f2f0848f8cc59edd9fc76134593bebef2de31S3Bucket3FFC6583": Object {
      "Description": "S3 bucket for asset \\"ac7f5d00801594308615088ac65f2f0848f8cc59edd9fc76134593bebef2de31\\"",
      "Type": "String",
    },
    "AssetParametersac7f5d00801594308615088ac65f2f0848f8cc59edd9fc76134593bebef2de31S3VersionKey4932657D": Object {
      "Description": "S3 key for asset version \\"ac7f5d00801594308615088ac65f2f0848f8cc59edd9fc76134593bebef2de31\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691": Object {
      "DependsOn": Array [
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "exports.handler = (event, context) => {
    // eslint-disable-next-line @typescript-eslint/no-require-imports, import/no-extraneous-dependencies
    const s3 = new (require('aws-sdk').S3)();
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const https = require('https');
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const url = require('url');
    log(JSON.stringify(event, undefined, 2));
    const props = event.ResourceProperties;
    if (event.RequestType === 'Delete') {
        props.NotificationConfiguration = {}; // this is how you clean out notifications
    }
    const req = {
        Bucket: props.BucketName,
        NotificationConfiguration: props.NotificationConfiguration,
    };
    return s3.putBucketNotificationConfiguration(req, (err, data) => {
        log({ err, data });
        if (err) {
            return submitResponse('FAILED', err.message + \`\\\\nMore information in CloudWatch Log Stream: \${context.logStreamName}\`);
        }
        else {
            return submitResponse('SUCCESS');
        }
    });
    function log(obj) {
        console.error(event.RequestId, event.StackId, event.LogicalResourceId, obj);
    }
    // eslint-disable-next-line max-len
    // adapted from https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-code.html#cfn-lambda-function-code-cfnresponsemodule
    // to allow sending an error message as a reason.
    function submitResponse(responseStatus, reason) {
        const responseBody = JSON.stringify({
            Status: responseStatus,
            Reason: reason || 'See the details in CloudWatch Log Stream: ' + context.logStreamName,
            PhysicalResourceId: event.PhysicalResourceId || event.LogicalResourceId,
            StackId: event.StackId,
            RequestId: event.RequestId,
            LogicalResourceId: event.LogicalResourceId,
            NoEcho: false,
        });
        log({ responseBody });
        const parsedUrl = url.parse(event.ResponseURL);
        const options = {
            hostname: parsedUrl.hostname,
            port: 443,
            path: parsedUrl.path,
            method: 'PUT',
            headers: {
                'content-type': '',
                'content-length': responseBody.length,
            },
        };
        const request = https.request(options, (r) => {
            log({ statusCode: r.statusCode, statusMessage: r.statusMessage });
            context.done();
        });
        request.on('error', (error) => {
            log({ sendError: error });
            context.done();
        });
        request.write(responseBody);
        request.end();
    }
};",
        },
        "Description": "AWS CloudFormation handler for \\"Custom::S3BucketNotifications\\" resources (@aws-cdk/aws-s3)",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:PutBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "Roles": Array [
          Object {
            "Ref": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "FfmpegLayer1A967C48": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "python3.8",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters819e1c802f7a3ee062b88ee47bec19d4f4f9c0e9475221ce3d678e85581305a5S3Bucket181207DA",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters819e1c802f7a3ee062b88ee47bec19d4f4f9c0e9475221ce3d678e85581305a5S3VersionKey842E1070",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters819e1c802f7a3ee062b88ee47bec19d4f4f9c0e9475221ce3d678e85581305a5S3VersionKey842E1070",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "Lambda layer for ffmpeg",
        "LicenseInfo": "GPLv3",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "InputVideoBucket411CB768": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "InputVideoBucketAllowBucketNotificationsToTestStackVideoProcessingFnED89CD5F1EAC38D3": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "VideoProcessingFn0FC68B62",
            "Arn",
          ],
        },
        "Principal": "s3.amazonaws.com",
        "SourceAccount": Object {
          "Ref": "AWS::AccountId",
        },
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "InputVideoBucket411CB768",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "InputVideoBucketNotifications15DA4C12": Object {
      "DependsOn": Array [
        "InputVideoBucketAllowBucketNotificationsToTestStackVideoProcessingFnED89CD5F1EAC38D3",
      ],
      "Properties": Object {
        "BucketName": Object {
          "Ref": "InputVideoBucket411CB768",
        },
        "NotificationConfiguration": Object {
          "LambdaFunctionConfigurations": Array [
            Object {
              "Events": Array [
                "s3:ObjectCreated:Put",
              ],
              "Filter": Object {
                "Key": Object {
                  "FilterRules": Array [
                    Object {
                      "Name": "suffix",
                      "Value": ".mp4",
                    },
                  ],
                },
              },
              "LambdaFunctionArn": Object {
                "Fn::GetAtt": Array [
                  "VideoProcessingFn0FC68B62",
                  "Arn",
                ],
              },
            },
          ],
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "BucketNotificationsHandler050a0587b7544547bf325f094a3db8347ECC3691",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3BucketNotifications",
    },
    "OutputVideoBucket793FE4FF": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "VideoProcessingFn0FC68B62": Object {
      "DependsOn": Array [
        "VideoProcessingFnServiceRoleDefaultPolicyEAE90BCB",
        "VideoProcessingFnServiceRole17162B39",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersac7f5d00801594308615088ac65f2f0848f8cc59edd9fc76134593bebef2de31S3Bucket3FFC6583",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersac7f5d00801594308615088ac65f2f0848f8cc59edd9fc76134593bebef2de31S3VersionKey4932657D",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersac7f5d00801594308615088ac65f2f0848f8cc59edd9fc76134593bebef2de31S3VersionKey4932657D",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "S3_DESTINATION_BUCKET": Object {
              "Ref": "OutputVideoBucket793FE4FF",
            },
          },
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "FfmpegLayer1A967C48",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "VideoProcessingFnServiceRole17162B39",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
      },
      "Type": "AWS::Lambda::Function",
    },
    "VideoProcessingFnServiceRole17162B39": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "VideoProcessingFnServiceRoleDefaultPolicyEAE90BCB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "InputVideoBucket411CB768",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "InputVideoBucket411CB768",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "OutputVideoBucket793FE4FF",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "OutputVideoBucket793FE4FF",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "VideoProcessingFnServiceRoleDefaultPolicyEAE90BCB",
        "Roles": Array [
          Object {
            "Ref": "VideoProcessingFnServiceRole17162B39",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
